Feature add rspec

mterada1228 opened the pull request:
以下のRSpecを追加
- spec/models/race_spec.rb
- spec/requests/horses_spec.rb
- spec/requests/races_spec.rb

その他の修正は微修正

mterada1228 committed:
既存のコードに対するUnitテストを追加

mterada1228 committed:
add races request spec

mterada1228 committed:
merge commit by M.Terada

mterada1228 committed:
add race model spec 途中

mterada1228 committed:
add prmary key

mterada1228 committed:
add race spec

mterada1228 committed:
Merge branch 'feature-add-rspec' of github.com:mterada1228/keiba_expectation_app into feature-add-rspec

mterada1228 committed:
add race model spec

mterada1228 committed:
Merge branch 'develop' into feature-add-rspec

ershn commented on Gemfile:
```diff
@@ -35,6 +35,9 @@ group :development, :test do
   gem 'byebug', platforms: [:mri, :mingw, :x64_mingw]
   gem 'rspec-rails', '~> 3.6.0'
+  gem "factory_girl_rails", "~> 4.8.0"
+  gem 'faker', :git => 'https://github.com/faker-ruby/faker.git', :branch => 'master'
```
github の master ブランチのものをそのまま使う理由がなければ、普通は rubygem のものを使うべきです。

faker だと、以下の方が正しいとおもいます。

```ruby
gem 'faker', '~> 2.14'
```

https://rubygems.org/gems/faker/versions/2.14.0

mterada1228 replied:
GemFIleを修正して、ご指摘いただいた形に修正いたしました。

ershn replied:
OK

ershn commented on Gemfile:
```diff
@@ -60,3 +63,6 @@ gem 'tzinfo-data', platforms: [:mingw, :mswin, :x64_mingw, :jruby]
 gem 'seed_dump'
+
+# グラフツール
+gem "chartkick"
```
Gemfile では基本的に`~>`でバージョンを全部指定するか、何も指定しないかが一般的だと思います。

両方のやり方にメリットとデメリットがあるので、ご確認ください。

mterada1228 replied:
全てバージョンの指定ありに修正いたしました。

ershn replied:
OK

ershn commented on app/controllers/races_controller.rb:
```diff
@@ -1,12 +1,13 @@
-    @race_horses = @race.race_horses.sort do
-      |a,b| a[:horse_number].to_i <=> b[:horse_number].to_i
-    end 
+    @race = Race.find_by_id(params[:id])
```
`find`はプライマリーキーで探すので、ここで`find_by_id`を使う必要がないと思います。

mterada1228 replied:
`find`に修正いたしました。

ershn replied:
OK

ershn commented on app/helpers/races_helper.rb:
```diff
@@ -7,9 +7,9 @@ def race_description_row1(race)
+    row2.reject(&:blank?).join(' ')
   end
 
   def race_description_row3(race)
```
row1, 2, 3 はその名前だと、何を指してるのかがわからないので、表現しているデータを表す名前にした方がいいと思います。

mterada1228 replied:
それぞれ以下のように修正致しました。
+ `row1` => `race_schedule`
+ `row2` => `race_condition`
+ `row3` => `race_prizes`

ershn replied:
OK

ershn commented on db/schema.rb:
```diff
@@ -33,22 +32,23 @@
 
-  create_table "horses", options: "ENGINE=InnoDB DEFAULT CHARSET=utf8", force: :cascade do |t|
+  create_table "horses", id: false, force: :cascade do |t|
+    t.integer "id", null: false
```
このテーブルにはプライマリーキーがないですが、検索が遅くなるのと、各レコードを特定するためのユニークな値がないということになるので、極力避けるべきだと思います。

mterada1228 replied:
`primary_key`を明示的に指定いたしました。
また`schema.rb`上での表示のされ方は2パターンになっております。
1. 複合キーの場合
     +  (例) horse_race_results : `primary_key: ["horse_id", "race_result_id"]`
2. 単一キーの場合
     + (例) horses : `id: :bigint`

ershn replied:
OK

ershn commented on app/models/horse.rb:
```diff
@@ -1,4 +1,9 @@
+  has_many :horse_race_results
+  has_many :race_horses
+
+  self.primary_key = :id
```
プライマリーキーは id としてテーブル定義で指定していれば、ここで再度定義する必要がないかと思います。

mterada1228 replied:
定義を削除いたしました。

ershn replied:
OK

ershn commented on app/models/race.rb:
```diff
@@ -1,4 +1,8 @@
   has_many :race_horses
-end
 
+  self.primary_key = :id
```
> プライマリーキーは id としてテーブル定義で指定していれば、ここで再度定義する必要がないかと思います。

mterada1228 replied:
定義を削除しました

ershn replied:
OK

ershn reviewed (changes requested):
全部ではないですが、レビューをしましたので、ご確認ください。

mterada1228 committed:
commit by M.Terada

mterada1228 committed:
Merge branch 'feature-add-rspec' of github.com:mterada1228/keiba_expectation_app into feature-add-rspec

mterada1228 commented:
ここまで頂いた指摘について対応いたしましたので、改めてご確認をお願い致します。

mterada1228 committed:
指摘対応

mterada1228 committed:
Merge branch 'feature-add-raceResult-scatter' into feature-add-rspec

mterada1228 committed:
feature-add-raceResult-scatter ブランチをマージしたことによる修正を行う

mterada1228 commented:
`feature-add-raceResult-scatter` ブランチをマージ致しました。
改めてご確認をお願い致します。

ershn commented on Gemfile:
```diff
@@ -20,43 +20,50 @@ gem 'webpacker', '~> 4.0'
-# gem 'image_processing', '~> 1.2'
-
 # Reduces boot times through caching; required in config/boot.rb
 gem 'bootsnap', '>= 1.4.2', require: false
```
`>=`のみだと、それ以上のバージョンで何でもオッケーという意味になるので、アップデートでメージャーバージョンが変わってしまって、コードがそのままバグってしまうようになるかもしれません。

https://semver.org/

それを避けるために、さらに`~>`で許容されてるバージョンを指定する必要があります。

mterada1228 replied:
`Gemfile`を修正し全て`~>`に変更しました。

ershn replied:
OK

ershn commented on db/schema.rb:
```diff
@@ -10,83 +10,83 @@
 
-  create_table "horses", options: "ENGINE=InnoDB DEFAULT CHARSET=utf8", force: :cascade do |t|
-    t.string "name"
+  create_table "horses", id: :bigint, default: nil, options: "ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_0900_ai_ci", force: :cascade do |t|
```
こちらの`default: nil`が必要かどうかは確認をお願いします。

mterada1228 replied:
DBのテーブル定義上、`null`が不可になっています。
明示的に `default:nil`が表示されてしまっていますが、
実際には`null`が入ることはないので問題ないと判断しております。

![image](https://user-images.githubusercontent.com/49284339/101431646-a693cf80-394a-11eb-84cf-f1950fdf73e9.png)

ershn replied:
OK

ershn commented on app/controllers/races_controller.rb:
```diff
@@ -1,12 +1,13 @@
-    @race_horses = @race.race_horses.sort do
-      |a,b| a[:horse_number].to_i <=> b[:horse_number].to_i
-    end 
+    @race_horses = @race.race_horses_sorted_by_horse_number
```
以下のようにでも書けると思います。
```
@race.race_horses.order('horse_number')
```

mterada1228 replied:
以下の対応を実施しました。
1. `app/controllers/races_controller.rb` にご指摘いただいた内容の処理を追加
2. `app/models/race.rb` から`race_horses_sorted_by_horse_number`のメソッドを削除

ershn replied:
OK

ershn commented on config/database.yml:
```diff
@@ -23,7 +16,10 @@ development:
+  host: localhost
 
 production:
   url: postgres://ieyrutiplidged:efb177262cbd2360942202d0b9d2c91dbff9c2a346dd6ae464f4753cf5d857eb@ec2-54-158-222-248.compute-1.amazonaws.com:5432/d9c9qim5bhf600
```
testとdevelopmentではあまり問題にならないが、誰でも閲覧できちゃうので、このファイルでパスワードは管理しないべきです。

ローカルにパスワードを管理するや、別な、パスワードが公開されない形で管理したほうが安全です。

mterada1228 replied:
`config/credentials.yml.enc` にパスワードを記載し、
`database.yml`からは`credentials.yml`の内容を参照するように変更しました。

`config/credentials.yml.enc`  の内容を確認する場合は、別途 `master.key` を slackにて送付いたします。

ershn replied:
OK

ershn commented on config/webpacker.yml:
```diff
@@ -53,7 +53,7 @@ development:
 
   # Verifies that correct packages and versions are installed by inspecting package.json, yarn.lock, and node_modules
-  check_yarn_integrity: true
+  check_yarn_integrity: false
```
開発でこちらのほうが便利という話でしたが、セキュリティー的には危険な設定かと思いますので、無効化せずに開発時の他の対策を考えたほうが安全だと思います。

mterada1228 replied:
`false` => `true` に変更いたしました。

ershn replied:
OK

ershn reviewed (changes requested):
差分にコメントを残しましたので、ご確認ください。

追加で以下も対応をお願いします。

`db/seeds.rb`に DB に入れるデータがすべて入ってるとのことですが、そのファイルがでかいので、リポジトリのデータ領域を無駄に使ってしまうのと、バージョン管理をする必要がないものなので、データはそれぞれの環境でスクリプトを通してライブで DBに入れるべきだと思います。

mterada1228 committed:
指摘対応
- Gemfile のバージョン指定を >= から ~> に変更
- races#show にhorse modelから競走馬のソート処理を簡略化して移動
- database.yml の password を credentials.yml 管理に変更
- webpacker.yml の check_yarn_integrity を true に変更
- seeds.rb を削除

mterada1228 commented:
レビューありがとうございました。
指摘対応を行いましたのでレビューをお願い致します。

ershn commented on Gemfile:
```diff
@@ -5,49 +5,41 @@ ruby '2.5.3'
 gem 'puma', '~> 4.1'
 # Use SCSS for stylesheets
-gem 'sass-rails', '>= 6'
+gem 'sass-rails', '~> 6'
```
バージョンナンバーは一つしか時は`~>`は`>=`と同じ意味を持つので、ナンバーは２つ以上が必要です。

mterada1228 replied:
1桁しかバージョンの指定のないgemは、2桁以上で指定するように変更しております。

ershn replied:
OK

ershn commented on spec/models/race_spec.rb:
```diff
@@ -1,5 +1,4 @@
 require 'rails_helper'
```
こちらの`require`は裏で自動的に行われて、明示的に書く必要はないかもしれません。

mterada1228 replied:
以下の2つの手順で修正を実施いたしました。
1. `.rspec` に以下を記載
  ```
      --require rails_helper
  ```
2. `./spec/*`より `require` 文を削除

ershn replied:
OK

ershn commented on spec/models/race_spec.rb:
```diff
@@ -1,5 +1,4 @@
 
-describe Race do
-  pending "add some examples to (or delete) #{__FILE__}"
+describe Race, type: :model do
```
`type: :model`と`type: :request`の指定は必要ないと思います。

mterada1228 replied:
`./spec/* ` のファイルより、 `type: `の指定は削除しました。

ershn replied:
OK

ershn commented on spec/requests/races_spec.rb:
```diff
@@ -0,0 +1,33 @@
+      # インスタンス変数を取得する
+      race_horses = controller.instance_variable_get('@race_horses')
+      expect(race_horses.pluck(:horse_number)).to eq((1..18).to_a)
+    end
```
以上の２つのテストで実行されてる処理（`get race_path(race)`）は一緒なので、一つのテストに`expect`を統合するのがいいかと思います。

そうすれば、無駄な処理を避けて、テスト全体の実行が早く終わるようになります。

mterada1228 replied:
`リクエストに正常に応答できること` と `レースに出走する競走馬が馬番の昇順でソートされていること` の `it` は一つにまとめました。

ershn replied:
OK

ershn commented on spec/requests/races_spec.rb:
```diff
@@ -0,0 +1,33 @@
+  describe 'Get#index' do
+    it 'リクエストに正常に応答できること' do
+      get races_path
+      expect(response).to be_successful
```
レスポンスが成功しただけではなく、html に出力されたデータも期待通りのものとなってるのもテストしたほうが安全だと思います。

mterada1228 replied:
フィーチャースペックに HTML のテストを追加しております。
- `spec/features/races_spec.rb`
- `spec/features/horses_spec.rb`

ershn replied:
OK

mterada1228 committed:
指摘対応
- Spec で毎回 require 'rails_helper' しなくていいように .rspec を修正
- gem のバージョン指定を修正
- horses#show で horse_race_result#dateの降順でレース結果が表示されるように修正
- horses の feature スペックを追加
- `type: ` の不必要な記載を削除

mterada1228 commented:
@ershn レビューありがとうございます。
修正を実施致しましたので、改めてご確認をお願い致します。

ershn commented:
`spec/requests/`配下のものは`spec/features/`のものと重複してるので、削除をお願いします。

ershn commented:
`horses_controller.rb`内で`Horse.includes(:race_results).find(params[:id])`の`includes`は不要だと思います。

ershn commented:
`try`は`param.try { |param| }`のようにも使えますので、コードを省略できます。

mterada1228 committed:
指摘対応
